### DISCUSSION TITLE
Fail to run the executable of ariadne packaged by pyinstaller: ModuleNotFoundError: No module named 'percol.cli'.

### DISCUSSION THREAD

**Discussion by hongyi-zhao** [pyinstaller/pyinstaller#6493] [PyInstaller]
On Ubuntu 20.04.3 LTS, I packaged [ariadne](https://github.com/hongyi-zhao/ariadne) into a stand-alone executable with the following steps sucessfully:

```
$ pyenv shell 3.9.1
$ pyenv virtualenv --system-site-packages test
$ pyenv shell test
$ pip install -u pip pyinstaller
$ pip list |grep pyinstaller
pyinstaller                       4.7
pyinstaller-hooks-contrib         2021.5
$ git clone https://github.com/hongyi-zhao/ariadne.git ariadne.git
$ cd ariadne.git
$ pyinstaller bin/percol
```
The above process completed successfully, but when I tried to run the generated executable, the following error occurred:

```
$ ./dist/percol/percol 
Traceback (most recent call last):
  File "percol", line 12, in <module>
ModuleNotFoundError: No module named 'percol.cli'
[2239827] Failed to execute script 'percol' due to unhandled exception!
```
Any hints for fixing this problem?

Regards,
HZ

**Comment by rokm** [comment]
Add `--hiddenimport percol.cli` to your pyinstaller command. If there are other modules from `percol` missing afterwards, try `--collect-submodules percol` instead.

**Reply by hongyi-zhao** [reply]
Thank you very much for your guidance, now I tried the following:

`$ pyinstaller --hiddenimport percol.cli --collect-submodules percol bin/percol `

This time, the following message appeared:

```
$ ./dist/percol/percol 
                                    __
        ____  ___  ______________  / /
       / __ \/ _ \/ ___/ ___/ __ \/ /
      / /_/ /  __/ /  / /__/ /_/ / /
     / .___/\___/_/   \___/\____/_/
    /_/
                                 0.2.1 

You did not give any inputs to percol. Check following typical usages and try again.

(1) Giving a filename,

 $ percol /var/log/syslog

(2) or specifying a redirection.

 $ ps aux | percol

Traceback (most recent call last):
  File "percol", line 14, in <module>
  File "percol/cli.py", line 209, in main
    def main():
  File "percol/cli.py", line 177, in exit_program
    
NameError: name 'exit' is not defined
[3138212] Failed to execute script 'percol' due to unhandled exception!
```
It seems to be close to success - how to solve these remaining problems in the next step?

Regards,
HZ

**Reply by rokm** [reply]
Probably easiest way is to fix `percol` to use `sys.exit()` instead of `exit()` (in percol/cli.py, line 177). Or perhaps update to their git master branch, [which already has this fixed](https://github.com/mooz/percol/blob/4b28037e328da3d0fe8165c11b800cbaddcb525e/percol/cli.py#L182) - the last release (0.2.1) is rather old, anyway.

**Reply by hongyi-zhao** [reply]
> Probably easiest way is to fix `percol` to use `sys.exit()` instead of `exit()` (in percol/cli.py, line 177).

Sorry, but I can't find the corresponding culprit code, so don't know how to fix it and try again. 

Furthermore, based on the comparison list of Python packaging and distribution tools [here](https://pyoxidizer.readthedocs.io/en/stable/pyoxidizer_comparisons.html), I find the following methods work without the problem discussed here:

```
#https://stackoverflow.com/questions/63107313/is-there-an-alternative-to-pyinstaller-for-python-3-8
#https://pyoxidizer.readthedocs.io/en/stable/pyoxidizer_comparisons.html
$ pip install --upgrade cx_Freeze --pre
# The generated executable is dist/perpol
$ cxfreeze -c bin/percol --packages curses,cmd --target-dir dist
# or using the following method:
# The generated executable is perpol.bin
$ pip  install -U nuitka
$ nuitka3 --follow-stdlib --follow-imports --static-libpython=no bin/percol
#If using on the same machine, the following is enough:
$ nuitka3 --static-libpython=no bin/percol
```

> Or perhaps update to their git master branch, [which already has this fixed](https://github.com/mooz/percol/blob/4b28037e328da3d0fe8165c11b800cbaddcb525e/percol/cli.py#L182) - the last release (0.2.1) is rather old, anyway.

Tried with the percol git master branch, but still failed, as follows:
```
$ git clone https://github.com/mooz/percol.git percol.git
$ cd percol.git
$ git log -1
commit 4b28037e328da3d0fe8165c11b800cbaddcb525e (HEAD -> master, origin/master, origin/HEAD)
Merge: 4aa06de a281f2b
Author: Masafumi Oyamada <stillpedant@gmail.com>
Date:   Wed Jul 24 08:48:07 2019 +0900

    Merge pull request #102 from lost-theory/no_output
    
    add a no_output action and document basic use of the API

$ pyenv shell datasci
$ pyinstaller --hiddenimport percol.cli --collect-submodules percol bin/percol
$ ./dist/percol/percol 
                                    __
        ____  ___  ______________  / /
       / __ \/ _ \/ ___/ ___/ __ \/ /
      / /_/ /  __/ /  / /__/ /_/ / /
     / .___/\___/_/   \___/\____/_/
    /_/
                                 0.2.1 

You did not give any inputs to percol. Check following typical usages and try again.

(1) Giving a filename,

 $ percol /var/log/syslog

(2) or specifying a redirection.

 $ ps aux | percol

Traceback (most recent call last):
  File "percol", line 14, in <module>
  File "percol/cli.py", line 209, in main
    exit_program(error_message("Cannot read a file '" + filename + "'"),
  File "percol/cli.py", line 177, in exit_program
    def exit_program(msg = None, show_help = True):
NameError: name 'exit' is not defined
[2490825] Failed to execute script 'percol' due to unhandled exception!
```
OTOH, the percol used by [ariadne](https://github.com/hongyi-zhao/ariadne), as noted [here](https://github.com/gawells/ariadne#ariadne), uses a modified version of Masafumi Oyamada's interactive grep tool: percol (https://github.com/mooz/percol). So the official percol can't be used directly by [ariadne](https://github.com/hongyi-zhao/ariadne).

**Reply by rokm** [reply]
I think the problem is a combination of search paths and the (old) `percol` package being installed in your virtual environment.

Your entry-point script is `bin/percol`, so `bin` sub-directory of your working directory (the git repository checkout) is added to search path, but your working directory is not. So the vendored `percol` that is part of the checked-out repository is not found.

However, you probably have a `percol` package installed in the (virtual) environment you are using, so PyInstaller ends up collecting modules from that one. Which is probably an old 0.2.1. version that is still using `exit` instead of `sys.exit`, hence the error.

Try adding the current working directory to search path (using `--paths .`), i.e.,
```
pyinstaller --clean --noconfirm --collect-submodules percol --paths . bin/percol
```
(You can also do a `pip uninstall percol`, just in case).

**Reply by hongyi-zhao** [reply]
Thank you very much for your expert-level in-depth and systematic analysis. The following steps do the trick:
```
$ pyenv shell datasci
$ pip uninstall percol 
$ pyinstaller --clean --noconfirm --collect-submodules percol --paths . bin/percol
# or 
$ pyinstaller --clean --noconfirm --collect-submodules percol --paths . -F --distpath pyinstaller bin/percol
```
But I noticed that the one-file bundled executable with `-F` option has a long startup time and is therefore less efficient:
```
# This is the executable generated by:
# $ pyinstaller --clean --noconfirm --collect-submodules percol --paths . bin/percol
$ time ./dist/percol/percol 
                                    __
        ____  ___  ______________  / /
       / __ \/ _ \/ ___/ ___/ __ \/ /
      / /_/ /  __/ /  / /__/ /_/ / /
     / .___/\___/_/   \___/\____/_/
    /_/
                                 0.2.1 

You did not give any inputs to percol. Check following typical usages and try again.

(1) Giving a filename,

 $ percol /var/log/syslog

(2) or specifying a redirection.

 $ ps aux | percol


real	0m0.108s
user	0m0.084s
sys	0m0.027s

# This is the executable generated by:
# $ pyinstaller --clean --noconfirm --collect-submodules percol --paths . -F --distpath pyinstaller bin/percol
$ time ./pyinstaller/percol 
                                    __
        ____  ___  ______________  / /
       / __ \/ _ \/ ___/ ___/ __ \/ /
      / /_/ /  __/ /  / /__/ /_/ / /
     / .___/\___/_/   \___/\____/_/
    /_/
                                 0.2.1 

You did not give any inputs to percol. Check following typical usages and try again.

(1) Giving a filename,

 $ percol /var/log/syslog

(2) or specifying a redirection.

 $ ps aux | percol


real	0m0.352s
user	0m0.282s
sys	0m0.072s
```
Any hints for this phenomenon?

Regards,
HZ

**Reply by rokm** [reply]
> But I noticed that the one-file bundled executable with `-F` option has a long startup time and is therefore less efficient:
> Any hints for this phenomenon?

That's because in onefile mode, everything is collected into executable-embedded archive, which needs to be extracted to a temporary directory each time you run the program. So it is expected to have a longer startup time (which might be significantly longer, depending on the size of collected resources).

**Reply by hongyi-zhao** [reply]
> So it is expected to have a longer startup time (which might be significantly longer, depending on the size of collected resources).

If this time is too long, this method may not have practical value for specific application scenarios.

**Reply by rokm** [reply]
> If this time is too long, this method may not have practical value for specific application scenarios.

True, but on the other hand, nobody is forcing you to use it.

**Reply by hongyi-zhao** [reply]
[Here](https://github.com/Nuitka/Nuitka/issues/1375) is another related discussion, just for reference.
